src/simd_fnt.h

```diff
@@ -0,0 +1,555 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __QUAD_SIMD_FNT_H__
+#define __QUAD_SIMD_FNT_H__
+
+#include <x86intrin.h>
+
+namespace quadiron {
+namespace simd {
+
+/* ================= Vectorized Operations ================= */
+
+// butterfly CT with r == 1
+template <typename T>
+inline void BUTTERFLY_1(VecType* x, VecType* y, T q)
+{
+    VecType add = ADD_MOD(*x, *y, q);
+    *y = SUB_MOD(*x, *y, q);
+    *x = add;
+}
+
+// butterfly CT with r == q - 1
+template <typename T>
+inline void BUTTERFLY_2(VecType* x, VecType* y, T q)
+{
+    VecType add = ADD_MOD(*x, *y, q);
+    *x = SUB_MOD(*x, *y, q);
+    *y = add;
+}
+
+// butterfly CT with 1 < r < q - 1
+template <typename T>
+inline void BUTTERFLY_3(VecType c, VecType* x, VecType* y, T q)
+{
+    VecType z = MUL_MOD(c, *y, q);
+    *y = SUB_MOD(*x, z, q);
+    *x = ADD_MOD(*x, z, q);
+}
+
+template <typename T>
+inline void BUTTERFLY_CT(T rp1, VecType c, VecType* x, VecType* y, T q)
+{
+    if (rp1 == 2) {
+        BUTTERFLY_1(x, y, q);
+    } else if (rp1 < q) {
+        BUTTERFLY_3(c, x, y, q);
+    } else {
+        BUTTERFLY_2(x, y, q);
+    }
+}
+
+// butterfly GS w/ r = q - 1
+template <typename T>
+inline void BUTTERFLY_4(VecType* x, VecType* y, T q)
+{
+    VecType add = ADD_MOD(*x, *y, q);
+    *y = SUB_MOD(*y, *x, q);
+    *x = add;
+}
+
+// butterfly GS w/ 1 < r < q - 1
+// x = x + y mod q
+// y = z * (x - y) mod q
+template <typename T>
+inline void BUTTERFLY_5(VecType c, VecType* x, VecType* y, T q)
+{
+    VecType sub = SUB_MOD(*x, *y, q);
+    *x = ADD_MOD(*x, *y, q);
+    *y = MUL_MOD(c, sub, q);
+}
+
+template <typename T>
+inline void BUTTERFLY_GS(T rp1, VecType c, VecType* x, VecType* y, T q)
+{
+    if (rp1 == 2) {
+        BUTTERFLY_1(x, y, q);
+    } else if (rp1 < q) {
+        BUTTERFLY_5(c, x, y, q);
+    } else {
+        BUTTERFLY_4(x, y, q);
+    }
+}
+
+template <typename T>
+inline VecType BUTTERFLY_GS_SIMPLE(T rp1, VecType c, VecType x, T q)
+{
+    if (rp1 == 2) {
+        return x;
+    } else if (rp1 < q) {
+        return MUL_MOD(c, x, q);
+    } else {
+        return NEG_MOD(x, q);
+    }
+}
+
+/**
+ * Vectorized butterly CT step
+ *
+ * For each pair (P, Q) = (buf[i], buf[i + m]) for step = 2 * m and coef `r`
+ *      P = P + r * Q
+ *      Q = P - r * Q
+ *
+ * @param buf - working buffers
+ * @param r - coefficient
+ * @param start - index of buffer among `m` ones
+ * @param m - current group size
+ * @param step - next loop
+ * @param len - number of vectors per buffer
+ * @param card - modulo cardinal
+ */
+template <typename T>
+inline void butterfly_ct_step(
+    vec::Buffers<T>& buf,
+    T r,
+    unsigned start,
+    unsigned m,
+    unsigned step,
+    size_t len,
+    T card)
+{
+    if (len == 0) {
+        return;
+    }
+    const T rp1 = r + 1;
+    VecType c = SET1(r);
+
+    const size_t end = (len > 1) ? len - 1 : 0;
+    const unsigned bufs_nb = buf.get_n();
+    // #pragma omp parallel for
```

Comment:
Should be removed:
- we are not using OpenPM yet
- this is dead code (it's commented)
- it's not related to this PR

Same comment for all the other occurrences.